### ISSUE TITLE
Callback trigger another callback

### ISSUE THREAD

**Issue by satomhxl** [sysrepo/sysrepo#2961]
Hello, I got a problem when using the callback function.

First, I register two callbacks monitoring changes of different xpath, see below:

sr_module_change_subscribe(sess, "o-ran-usermgmt", "/o-ran-usermgmt:users/user/name", callback_a, NULL, 0, SR_SUBSCR_DEFAULT, subscription);
sr_module_change_subscribe(sess, "o-ran-usermgmt", "/o-ran-usermgmt:users/user/enabled", callback_b, NULL, 0, SR_SUBSCR_DEFAULT, subscription);

Then, I trigger callback_a, and in this callback, I call "sr_set_item" and "sr_apply_changes", to modify the value of an item whose xpath is "/o-ran-usermgmt:users/user/enabled", so I think this action will trigger callback_b.

But I found in gdb that after the execution of "sr_apply_change" in callback_a, the program just halt, and after several seconds, it returns a "time out" error. The full output is:

ERROR
        type:     application
        tag:      operation-failed
        severity: error
        message:  Callback event "rpc" with ID 31 processing timed out.

        type:     application
        tag:      operation-failed
        severity: error
        message:  User callback failed.

And in gdb, I didn't see the callback_b is called.
Can you give me some advice on how to solve this? I just want the "sr_set_item" in callback_a can be executed successfully.
Thanks very much!

**Comment by satomhxl** [comment]
Well, I found the program is halted at  sr_apply_changes --> sr_changes_notify_store --> sr_shmsub_change_notify_change --> sr_shmsub_notify_wait_wr.
Seems the function just wait the write lock until time runs out.
So the problem is that I can't call sr_apply_changes in a sysrepo callback caused by SR_EV_CHANGE event, because the write lock is already occupied?

**Comment by michalvasko** [comment]
What you described is a nice example of a dead lock. You are handling an event of `name` on `subscription`, which generates an event of `enabled` on `subscription`. However, `subscription` is still handling the former event so it cannot handle the latter one. The solution is simple, use 2 subscription structures, one for each event. Then they can generate events for each other.

**Comment by michalvasko** [comment]
> You mean the last parameter passed to sr_module_change_subscribe() should be different?

Yes.

> The result of sr_apply_changes() in callback_a is still timeout.

Ah, yes, that is probably right. There is another dead lock there then, when you are modifying module `o-ran-usermgmt` and you try to modify it again, in the callback. To prevent that, modify the data on `SR_EV_DONE` event, not on `SR_EV_CHANGE`.

**Comment by satomhxl** [comment]
Emmm... I did as you said, execute sr_set_item() and sr_apply_changes() in callback_a on SR_EV_DONE event, but the result of sr_apply_changes() is still timeout.

**Comment by michalvasko** [comment]
Are you sure you do not have any other callbacks subscribed? Please mention, again, all the details of this last use-case. Unless you are triggering the same callback that is being processed, it should work.

**Comment by Podendo** [comment]
Hi @michalvasko 
Just for clarifying:
So it is not possible to apply changes in one callback for  a node that is also used in subscription in another callback, in the same module,
i.e:
callback_A with own context_A,  priority 0, subscribed with default flags to xpath:  my_module:container_A/data_A
callback_B with own context_B, priority 5, subscribed with default flags to xpath:  my_module:container_B/data_B
Triggering callback_A: setting values for data_B with sr_apply_changes(...) at SR_EV_DONE leads to timeout expiration, and SR_EV_ABORT for callback_B?

**Comment by michalvasko** [comment]
No, on `SR_EV_DONE` it should not because by then `callback_A` should not be holding `my_module` lock.

**Comment by Podendo** [comment]
This is odd. At least for sysrepo v2.2.12 i still get timeout and callback failure.
By logs i can see that events are published for a whole module.
So when callback A is triggered and sr_apply_changes is called, i can see event publishing/processing for my module and each subscriber, but no callbacks are triggered).  At the point of publishing event change for callback A (verified by priority) i can see the timeout expiration and publishing event abort.
Right after, callback B is triggered with event abort. After event abort is processed, i can see the failure message from callback A caused by sr_apply_changes non-zero return value.

**Comment by michalvasko** [comment]
Sorry, I have looked at the code and it seems that if you want to modify data of the **same** module from a module_change callback, it will cause a dead-lock for both `SR_EV_CHANGE` and `SR_EV_DONE` events. However, you should be able to do this on `SR_EV_UPDATE` event.
